Add asymptotic confidence sequences to the monitoring user guide (#358)
EmertonData/glide: docs/user_guide/monitors.md

### Asymptotic Confidence Sequences

This time, the confidence sequence is built by exploiting the variance of each batch estimate, rather than only the fact that it lies in $[0, 1]$.

#### Setting

Each batch $t$ contributes a per-batch estimate $\hat{R}_t$ together with its own standard error $\hat{\sigma}_t$. This applies, for example, when $\hat{R}_t$ is a sample mean of the batch's labeled values, with standard error computed according to the Central Limit Theorem. As before, the running risk after $t$ batches is $\bar{R}_t = \frac{1}{t}\sum_{s=1}^{t}\hat{R}_s$, monitored against a user-fixed threshold $\tau$, and each $\hat{R}_s$ is assumed to be approximately Gaussian around the batch's own risk.

Define the **intrinsic time**

$$\nu_t = \sum_{s \le t} \hat{\sigma}_s^2,$$

the accumulated variances of the per-batch estimates.

#### From a Gaussian Wealth Process to an Anytime-Valid Bound

> The anytime-valid guarantee rests on the same two classical facts used throughout this guide. Markov's inequality states that a nonnegative random variable $W$ with $E[W] \le 1$ satisfies $\Pr(W \ge 1/\delta) \le \delta \, E[W] \le \delta$. Ville's inequality upgrades this to a nonnegative supermartingale $\{W_t\}_{t \ge 0}$ with $W_0 = 1$ yielding: $\Pr(\exists t \ge 1 : W_t \ge 1/\delta) \le \delta$, bounding the probability of *ever* crossing $1/\delta$ rather than the probability at one fixed time.

Reviewer: Est-ce qu'il faudrait pas faire une section tout en haut sur les fondamentaux (Markov, Ville) plutôt que d'attacher ça à la section Empirical Bernstein ?

Author: Fait

Read $W_t$ as the wealth of a gambler betting against $H_0$, the null hypothesis of no drift, starting with one unit of capital. For a betting rate $\lambda \in \mathbb{R}$, define

$$W_t(\lambda) = \exp\!\left(\lambda S_t - \frac{\lambda^2}{2}\nu_t\right),$$

where $S_t = \sum_{s \le t}(\hat{R}_s - c_s)$ is the cumulative deviation of the per-batch estimates from their predictable centers $c_s = \bar{R}_{s-1}$ (with $c_1 = \tau$ for the first batch). If the deviations $\hat{R}_s - c_s$ were exactly $\mathcal{N}(0, \hat{\sigma}_s^2)$-distributed, $W_t(\lambda)$ would be a supermartingale for every $\lambda$.

Rather than commit to one $\lambda$, mix over a **folded Gaussian** density of scale $\rho$: a Gaussian density restricted to $\lambda > 0$ and doubled so that it integrates to one. This is the natural conjugate mixing distribution for Gaussian increments [[5](#ref-5)]:

$$W_t = \int_0^\infty W_t(\lambda) \cdot \frac{2}{\sqrt{2\pi\rho^2}}\exp\!\left(-\frac{\lambda^2}{2\rho^2}\right) d\lambda.$$

> A mixture of nonnegative supermartingales, each starting at $1$, is itself a nonnegative supermartingale starting at $1$, so Ville's inequality applies for any choice of $\rho > 0$; the scale determines the tightness of downstream bounds. Carrying out the Gaussian integral and inverting $W_t \ge 1/\delta$ for the largest deviation still consistent with $H_0$, the same "boundary" idea used elsewhere in this guide but solvable here in closed form, gives the anytime-valid lower bound on the running risk after $t$ batches,

Reviewer: ```suggestion
A mixture of nonnegative supermartingales, each starting at $1$, is itself a nonnegative supermartingale starting at $1$, so Ville's inequality applies for any choice of $\rho > 0$; the scale determines the tightness of downstream bounds. Carrying out the Gaussian integral and inverting $W_t \ge 1/\delta$ for the largest deviation still consistent with $H_0$, we can obtainthe anytime-valid lower bound on the running risk after $t$ batches with a closed form,
```

Author: Appliqué

> $$L_t = \bar{R}_t - \sqrt{ \frac{2(\nu_t \rho^2 + 1)}{t^2 \rho^2} \log\!\left(1 + \frac{\sqrt{\nu_t \rho^2 + 1}}{2\delta}\right) }.$$

Reviewer: Je me pose une question ici: la différence L_T - R_t ne dépend que de nu_t, et donc pas de R_t. ça veut dire que la lower bound est juste translatée, mais indépendante de si R_t est proche de 0 ou pas par exemple. Normal ?

Author: Le détail important ici c'est qu'on utilise $\bar{R}_t$, la moyenne cumulée, pour définir la borne inférieure. Ainsi, une valeur "étonnante" de $R_t$ à un instant donné serait diluée dans l'historique.

This bound is exact if the per-batch deviations are truly Gaussian. In practice they are only approximately so that a further argument with a strong approximation is needed (see [[4](#ref-4)] for details). It shows that the partial sums of per-batch estimates stay close to those of a genuinely Gaussian process so that the same boundary remains valid for a sufficient number of batches. This makes the guarantee asymptotic rather than exact.

#### Tuning and Interpreting the Boundary

No anytime-valid boundary can be tight at every batch: tightening it at one horizon necessarily loosens it at others, so the scale $\rho$ controls where that tightness is spent. Choosing

$$\rho^2 = \frac{-2\log(2\delta) + \log\bigl(-2\log(2\delta) + 1\bigr)}{\nu_{t^\star}}$$

makes $L_t$ tightest at a user-chosen target batch $t^\star$. Note, however, that the penalty for choosing a different target than needed is generally mild in practice.

> The above boundary's width scales as $\sqrt{\nu_t \log \nu_t}/t$, shrinking with the actual precision of the per-batch estimates. Moreover, since $\nu_t$ accumulates each batch's own internal standard error rather than deviations between batches, a shift in the risk from one batch to the next only affects $\bar{R}_t$ and not $\nu_t$: the boundary is immune to drift-inflation.

Reviewer: Moreover... -> tout ça semble être comparaison implicite à empirical Bernstein, on peut l'enlever

Author: Fait

To compute predictable weights $\lambda_t$, the simplest option is to pool the full prior history of batches. The first batch has no predecessor, so it can use the neutral weight $\lambda_1 = 1$, which is itself trivially predictable.

#### Bound and Alarm Rule

With both additions in place, plugging the clipped $\hat{R}_t$ into the [Empirical-Bernstein Confidence Sequences](#empirical-bernstein-confidence-sequences) construction gives the anytime-valid lower bound $L_t = \bar{R}_t - u(V_t)/t$ defining the drift alarm rule as,

$$L_t > \tau,$$

now backed by the more sample-efficient PPI++ estimate, at the same single false-alarm budget $\delta$.

### Asymptotic Classical Risk Monitoring

The asymptotic confidence sequence above is applied here in the setting where the per-batch estimate comes from a batch of human-labeled samples.

#### Setting

Each batch $t$ contributes a set of human labels.

| Value | Present for | Description |
|---|---|---|
| $Y_{t,j}$ | All labeled samples in batch $t$ | Ground-truth label |

> Every batch is monitored relative to a user-fixed threshold $\tau$ and the metric is treated as a **risk** $R$, where lower is better; a performance metric is monitored by applying the same methodology to $1 - R$ instead of $R$.

Reviewer: Dropper la fin sur performance metric, on l'a déjà dit plein de fois dans ce user guide

Author: Fait

The per-batch risk estimate is the classical sample mean of the $n_s$ labels in batch $s$,

$$\hat{R}_s = \frac{1}{n_s}\sum_{j=1}^{n_s} Y_{s,j},$$

together with its standard error,

$$\hat{\sigma}_s = \sqrt{\frac{\widehat{\mathrm{Var}}(Y_s)}{n_s}}.$$

#### Bound and Alarm Rule

Plugging $\hat{R}_s$ and $\hat{\sigma}_s$ into the [Asymptotic Confidence Sequences](#asymptotic-confidence-sequences) gives the anytime-valid lower bound on the running risk,

> $$L_t = \bar{R}_t - \sqrt{ \frac{2(\nu_t \rho^2 + 1)}{t^2 \rho^2} \log\!\left(1 + \frac{\sqrt{\nu_t \rho^2 + 1}}{2\delta}\right) },$$

Reviewer: A-t-on vraiment besoin de répéter cette formule, qui est inchangée par rapport à la section confidence sequence ?

Author: L'idée était que cette parie reste auto-suffisante. Mais ce n'était déjà pas le cas à cause du rho ... Je viens d'enlever la formule.

| $Y_{t,j}$ | Labeled samples in batch $t$ only | Ground-truth label |

The per-batch estimate $\hat{R}_s$ is the PPI++ estimate on batch $s$ as in Empirical PPRM, together with its standard error $\hat{\sigma}_s$, obtained by applying the [Prediction-Powered Inference (PPI++)](estimators.md#prediction-powered-inference-ppi) variance formula within batch $s$. Unlike Empirical PPRM, $\hat{R}_s$ is used directly here, without clipping onto $[0, 1]$: the asymptotic construction does not require the estimate itself to lie in a bounded range, only that it is approximately Gaussian with a known standard error.

#### Predictable Power-Tuning

As in Empirical PPRM, the power-tuning weight $\lambda_t$ used to form $\hat{R}_t$ must be **predictable**: computed only from batches strictly earlier than $t$. The power-tuning factors are therefore computed using strictly earlier batches. The first batch has no predecessor, so it uses the neutral weight $\lambda_1 = 1$.

#### Bound and Alarm Rule

Plugging $\hat{R}_s$ and $\hat{\sigma}_s$ into the [Asymptotic Confidence Sequences](#asymptotic-confidence-sequences) construction gives the anytime-valid lower bound on the running risk,

> $$L_t = \bar{R}_t - \sqrt{ \frac{2(\nu_t \rho^2 + 1)}{t^2 \rho^2} \log\!\left(1 + \frac{\sqrt{\nu_t \rho^2 + 1}}{2\delta}\right) },$$

Reviewer: A-t-on vraiment besoin de répéter cette formule, qui est inchangée par rapport à la section confidence sequence ?

Author: L'idée était que cette parie reste auto-suffisante. Mais ce n'était déjà pas le cas à cause du rho ... Je viens d'enlever la formule.
